doc/design/zarr-support-2.md 

```diff
@@ -0,0 +1,94 @@
+  Ask to upload a batch of subfiles as part of the zarr file.
+  No more batch uploads are permitted until this one is completed or aborted.
+  Requires a list of file paths and ETags (md5 checksums)
+  The file paths may include already uploaded files; this is how updates are done.
+  Returns a list of presigned upload URLs
+* POST /api/zarr_files/{zarr_id}/batch/complete/
+  Completes an upload of a batch of subfiles.
+  Fail if any of the checksums do not match.
+  Also, recursively update the tree hash for every parent node for each child.
+* DELETE /api/zarr_files/{zarr_id}/batch/
+  Cancels a batch upload.
+  Any files already uploaded are deleted.
+  A new batch upload can then be started.
+* DELETE /api/zarr_files/{zarr_id}/batch/delete/
+  Deletes subfiles from S3, and updates the treehash accordingly.
+  Requires a list of file paths
+
+* POST /api/dandisets/{...}/versions/{...}/assets/zarr/{zarr_id}/
+  Creates a new Asset that points to this zarr data.
+  Return an asset ID
+
+## Upload flow
+
+TODO the only gotcha is that published zarr files can't be uploaded to/updated/deleted from
+The slowest API request in the upload flow is the `.../batch/complete/` operation, since it needs to update several `.checksum` files in S3.
+I recommend that the CLI dynamically adjust batch size to keep the `.../batch/complete/` operation under 30s.
+
+## Modification flow
+
+TODO this is basically the same as the upload flow, but with deletes allowed.
+
+## Hashing scheme
+
+To avoid having to store a checksum for every directory in every zarr file, we will delegate the storage of the checksums to S3.
+Update times will be substantially slower since each level of the tree needs to be updated.
+Storage costs will be offloaded to S3, and the treehash will be conveniently accessible to clients.
+
+Let's say the zarr files are kept under `/zarr/{zarr_id}/...` in S3.
+A subfile `1/2/3/4.txt` of zarr file `foobar` would be kept at `/zarr/foobar/1/2/3/4.txt`.
+The ETag can be trivially determined by a `HEAD` request to that S3 object.
+
+After this subfile is uploaded, the API server creates a file `/zarr_checksums/foobar/1/2/3/.checksum` if it doesn't already exist (note the altered prefix).
+This file contains a list of all the files/directories currently in the directory, their checksums, and an aggregated checksum of the entire directory contents.
+If `.checksum` already exists, it is updated with the new file+ETag and the final checksum is updated.
+
+The API server then recursively travels up another level and creates `/zarr_checksums/foobar/1/2/.checksum`.
+This `.checksum` will contain an entry for the directory `3` and the final checksum from `3/.checksum`.
+This recursion continues all the way to the top, yielding a `/zarr/checksums/foobar/.checksum` file which contains an aggregated checksum for the entire zarr file.
+
+### .checksum file format
+
+TODO
+TODO include sha256 as it is calculated? are we even calculating sha256 for zarr files?
```

Comment:
good question: might become a huge load on the backend server to send 1mil queries for offband checksuming tiny files... probably we can get away and just compute md5/etag?

Reply:
Yeah, my thought was that sha256 might come in handy for a .nwb file, but taking the sha256 of a zarr file doesn't make any sense to me. If we were to sha256 each subfile we could throw it into the .checksum file somewhere, but actually cascading it upwards seems redundant with the ETag based treehash we are already computing.